dali/operators/reader/loader/numpy_loader.cc

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2017-2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <dirent.h>
+#include <errno.h>
+#include <memory>
+
+#include "dali/core/common.h"
+#include "dali/operators/reader/loader/numpy_loader.h"
+#include "dali/util/file.h"
+#include "dali/operators/reader/loader/utils.h"
+
+
+namespace dali {
+TypeInfo TypeFromNumpyStr(const std::string &format) {
+  if (format == "i8")
+    return TypeInfo::Create<int64_t>();
+  else if (format == "i4")
+    return TypeInfo::Create<int32_t>();
+  else if (format == "f4")
+    return TypeInfo::Create<float>();
+  else if (format == "f8")
+    return TypeInfo::Create<double>();
+  else
+    return TypeInfo();
+}
+
+std::unique_ptr<FileStream> NumpyLoader::ParseHeader(std::unique_ptr<FileStream> file,
+                                                     NumpyParseTarget& target) {
+  // check if the file is actually a numpy file
+  uint8_t* token = new uint8_t[10];
+  int64_t nread = file->Read(token, 10);
+  DALI_ENFORCE(nread == 10, "Can not read header.");
+
+  // check if heqder is too short
+  char* chartoken = reinterpret_cast<char*>(token);
+  DALI_ENFORCE(std::string(chartoken).find("NUMPY") != std::string::npos,
+               "File is not a numpy file.");
+
+  // extract header length
+  uint16_t header_len = 0;
+  memcpy(&header_len, &token[8], 2);
+  delete [] token;
+
+  // read header: the offset is a magic number
+  int64 offset = (6+1+1+2);
+  token = new uint8_t[header_len];
+  chartoken = reinterpret_cast<char*>(token);
```

Comment:
Ugly

Reply:
I was thinking about that. I think this is the only way. If I allocate a big enough string and pass that to the read function, I would need to cast that to uint8_t. But that is more dangerous. Afaik, casting from uint8_t* to char* is fine, the other way round is not.